src/lib/core/CASEAuthTag.h

```diff
@@ -84,6 +87,16 @@ struct CATValues
         }
         return CHIP_NO_ERROR;
     }
+
+private:
+    static bool Cmp(const CASEAuthTag & a, const CASEAuthTag & b) { return a < b; }
+    void Sort() { Sorting::BubbleSort(this->values.begin(), this->size(), Cmp); }
+    CATValues Sorted() const
+    {
+        auto sorted = *this;
```

Comment:
The invariant at line 52/53 is broken by sorting:

```
    // All valid CAT values are always in the beginning of the array followed by kUndefinedCAT values.
    ReturnErrorCodeIf(cat == kUndefinedCAT, false);
```

Reply:
There appears to be nothing at all in this object that enforces this in the first place.

I think I will do a reverse sort in the constructor to make this _actually_ guaranteed.  That should cleanly solve everything at once.

Reply:
Ah, I see we can't do this because we are transcribing directly to and from certificates and ordering _does_ matter for the signature.  And this is apparently where the promise that kUndefinedCAT will appear at the end of the array is realized.

Also, the code seems to be pretty set on CATValues being an aggregate.

I'll stick to the original approach, but with reverse ordering for the temporary arrays, and audit to see how often we actually use the equality operator.  Hopefully not often.

Reply:
> This means that after sorting, iteration may fail and CheckSubjectAgainstCATs may iterate less entries (or none?)

Ah, OK.  So the thing is that the Sorted arrays are only temporary and only used for comparison.  And per the above, we can't sort in place anyway because inside the certificate, order _does_ matter.

For this reason, this approach actually doesn't break anything.  Is there anything beyond making these methods private that I can do to address your concern?

Here is where we are using the comparison:

```
../src/transport/SecureSessionTable.h:114:43: error: no match for ‘operator==’ (operand types are ‘const chip::CATValues’ and ‘const chip::CATValues’)
  114 |                 oldSession->GetPeerCATs() == session->GetPeerCATs())
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~ ^~ ~~~~~~~~~~~~~~~~~~~~~~
      |                                        |                         |
      |                                        const chip::CATValues     const chip::CATValues
```

We don't compare these often, but where we do, the comparison is broken as I have claimed.  So a fix is needed.

Reply:
Collaborated on a solution, which is now merged into this branch.